flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java

```diff
@@ -0,0 +1,272 @@
+                        .build();
+
+        INNER_JOIN =
+                TableTestProgram.of("inner join", "test inner join")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a1 int", "b1 int")
+                                        .consumedBeforeRestore(
+                                                Row.of(1, 1),
+                                                Row.of(2, 2),
+                                                Row.of(3, 3),
+                                                Row.of(2, 2))
+                                        .consumedAfterRestore(Row.of(2, 2))
+                                        .build())
+                        .runSql("insert into MySink " + "SELECT a1, b1 FROM A JOIN B ON a1 = b1")
+                        .build();
+
+        JOIN_WITH_FILTER =
+                TableTestProgram.of("join-with-filter", "test join with filter")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a3 varchar", "b4 varchar")
+                                        .consumedBeforeRestore(
+                                                Row.of("Hi", "Hallo"),
+                                                Row.of("Hi", "Hallo Welt wie gehts"))
+                                        .consumedAfterRestore(new Row[] {})
+                                        .build())
+                        .runSql(
+                                "insert into MySink "
+                                        + "SELECT a3, b4 FROM A, B where a2 = b2 and a2 < 2")
+                        .build();
+
+        INNER_JOIN_WITH_DUPLICATE_KEY =
+                TableTestProgram.of(
+                                "inner-join-with-duplicate-key", "inner join with duplicate key")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a1 int", "b1 int")
+                                        .consumedBeforeRestore(Row.of(2, 2))
+                                        .consumedAfterRestore(new Row[] {})
+                                        .build())
+                        .runSql(
+                                "insert into MySink "
+                                        + "SELECT a1, b1 FROM A JOIN B ON a1 = b1 AND a1 = b3")
+                        .build();
+
+        INNER_JOIN_WITH_NON_EQUI_JOIN =
```

Comment:
How about we use the same data as for `JOIN_WITH_FILTER`? In the end it should end up with the same plan.

Reply:
The queries/plans are slightly different.

If we make put some light constraints on the input data, then the queries could be "closer" and share the same output.  

For instance, the two queries could become:

```
insert into MySink
SELECT a3, b4 FROM A, B where a2 = b2 and a2 < 2
```
and 
```
insert into MySink
SELECT a3, b4 FROM A, B where a2 = b2 and a2 > b2
```

That alignment would also use words instead of numbers (AKA, this would line up with the other feedback as well).

Does that sound like a plan?

Reply:
I think it's the same situation as here: https://github.com/apache/flink/pull/23680#discussion_r1387873562

I understand the only difference to be how we specify the join condition. If that's the case we can just use the same input/output.